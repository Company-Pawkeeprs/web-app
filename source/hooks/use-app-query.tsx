import {
    useQuery,
    type QueryKey,
    type UseQueryOptions,
} from '@tanstack/react-query'
import type { AxiosResponse } from 'axios'
import { useAppSelector } from '~/store/hooks'

export type Fn<T> = () => Promise<AxiosResponse<T>>

const useAppQuery = <T,>(
    key: QueryKey,
    fn: Fn<T>,
    options?: Omit<UseQueryOptions<T>, 'queryKey'>,
) => {
    const { user } = useAppSelector((state) => state.Profile)

    const Fn = () => fn?.()

    return useQuery<T>({
        queryKey: [user?.email, ...key],
        queryFn: async () => {
            const res = await Fn()
            return res.data
        },
        ...options,
    })
}

export default useAppQuery
