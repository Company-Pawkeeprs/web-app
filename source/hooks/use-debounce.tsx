import { useState } from 'react'

export default function useDebounce() {
    const [typingTimeout, setTypingTimeout] = useState<NodeJS.Timeout>()

    const debounce = async (func: () => any, wait = 1000) => {
        clearTimeout(typingTimeout!)

        const timeout = setTimeout(() => {
            func()
            // eslint-disable-next-line testing-library/await-async-utils
        }, wait)

        setTypingTimeout(timeout)
    }

    return debounce
}
