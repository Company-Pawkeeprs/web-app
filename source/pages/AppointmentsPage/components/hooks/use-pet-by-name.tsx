import { useEffect, useTransition } from "react"
import { Data } from "~/store/pets/types"

type UsePetByNameProps = {
    name?: string | null,
    pets: Data[],
    onChangeField: (field: string, value: any) => void
}

const usePetByName = ({ onChangeField, pets, name }: UsePetByNameProps) => {
    const [isPending, startTransition] = useTransition()

    useEffect(() => {
        if (!name) {
            return
        }

        const pet = pets.find(pet => pet.name === name)

        if (pet) return

        onChangeField('pet.id', '')
        onChangeField('pet.avatar', '')
    }, [name, pets, onChangeField])

    const onChangePet = (pet: Data) => {
        startTransition(() => {
            onChangeField('pet.id', pet.id)
            onChangeField('pet.avatar', pet.avatar)
            onChangeField('pet.breed', pet.breed)
            onChangeField('pet.species', pet.species)
            onChangeField('pet.bloodType', pet.bloodType)
            onChangeField('pet.gender', pet.gender)
            onChangeField('pet.dateOfBirth', pet.dateOfBirth)
        })
    }


    return {
        isPending,
        onChangePet
    }
}

export default usePetByName