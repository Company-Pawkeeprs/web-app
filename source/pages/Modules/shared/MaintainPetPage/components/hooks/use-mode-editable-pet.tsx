import { create } from 'zustand'
import type { ModeInput } from '~/Components/molecules/field-control/field-control'

type State = {
    mode: ModeInput
    onChangeMode: (mode: ModeInput) => void
    toggleMode: () => void
}

export const useModeEditablePet = create<State>((set) => ({
    mode: 'editable',
    onChangeMode: (mode) => set({ mode }),
    toggleMode: () =>
        set((state) => ({
            mode: state.mode === 'editable' ? 'readonly' : 'editable',
        })),
}))
