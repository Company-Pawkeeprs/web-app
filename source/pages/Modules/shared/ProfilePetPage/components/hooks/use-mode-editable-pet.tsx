import { create } from 'zustand'
import type { ModeInput } from '~/Components/molecules/field-control/field-control'

type State = {
    mode: ModeInput
    onChangeMode: (mode: ModeInput) => void
}

export const useModeEditablePet = create<State>((set) => ({
    mode: 'editable',
    onChangeMode: (mode) => set({ mode }),
}))
