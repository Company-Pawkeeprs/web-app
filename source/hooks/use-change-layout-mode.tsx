import { useCallback } from "react";
import { layoutModeTypes } from "~/Components/constants/layout";
import { useAppDispatch, useAppSelector } from "~/store/hooks";
import { changeLayoutMode } from "~/store/layouts/slice";

const userPreferenceMode = () => {
    let preferenceMode = localStorage.getItem("userPreferenceMode");

    if (!preferenceMode) {
        preferenceMode = "light";
        localStorage.setItem("userPreferenceMode", preferenceMode);
    } else {
        preferenceMode = preferenceMode === "dark" ? "light" : "dark";
        localStorage.setItem("userPreferenceMode", preferenceMode);
    }

    return preferenceMode;
};

const useChangeLayoutMode = () => {
    const dispatch = useAppDispatch();
    const layoutMode = useAppSelector((state) => state.Layout.layoutModeType);

    const onHandleChangeLayout = useCallback(() => {
        const mode =
            layoutMode === layoutModeTypes.LIGHT_MODE
                ? layoutModeTypes.DARK_MODE
                : layoutModeTypes.LIGHT_MODE;

        dispatch(changeLayoutMode(mode));
    }, [dispatch, layoutMode]);

    return {
        onHandleChangeLayout,
    };
};

export default useChangeLayoutMode;
