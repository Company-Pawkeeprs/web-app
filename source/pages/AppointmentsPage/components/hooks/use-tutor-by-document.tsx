import { useEffect, useState, useTransition } from "react"
import { useAppSelector } from "~/store/hooks"
import { Data as Pet } from "~/store/pets/types"

type UseTutorByDocumentProps = {
    document: string
    onChangeField: (field: string, value: any) => void
}

const useTutorByDocument = ({ document, onChangeField }: UseTutorByDocumentProps) => {

    const [isPending, startTransition] = useTransition()
    const [tutorExists, setTutorExists] = useState(false)
    const [petsOptions, setPetsOptions] = useState<(Pet & { value: string })[]>([])

    const { pets, tutors } = useAppSelector(state => ({
        pets: state.Pets.data,
        tutors: state.Tutor.data
    })
    )

    useEffect(() => {
        const documentNumber = document?.replace(/\D/g, '')

        if (!documentNumber || documentNumber.length < 11) {
            return
        }

        const tutor = tutors.find(tutor => tutor.document === documentNumber)

        if (!tutor) {
            setTutorExists(false)
            return
        }

        const petsOptions = pets
            .filter(pet => pet?.ownerEmergencyContact.document === documentNumber)
            .map(pet => ({
                ...pet,
                value: pet.id,
            }))

        startTransition(() => {
            setTutorExists(true)
            setPetsOptions(petsOptions)
            onChangeField('tutor.id', tutor?.id || '')
            onChangeField('tutor.phone', tutor?.phone || '')
            onChangeField('tutor.name', tutor?.name || '')
            onChangeField('tutor.email', tutor?.email || '')
            onChangeField('tutor.avatar', tutor?.avatar || '')
            tutor?.address && onChangeField('tutor.address', tutor?.address)
        })

    }, [document, tutors, pets, onChangeField])

    return {
        isPending,
        tutorExists,
        petsOptions
    }
}

export default useTutorByDocument