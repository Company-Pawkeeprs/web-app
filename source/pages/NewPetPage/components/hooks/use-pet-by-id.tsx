import { useCallback, useEffect, useState, useTransition } from "react"
import { useAppSelector } from "~/store/hooks"
import { IPet } from "~/types/pet"

type UsePetByNameProps = {
    id?: string | null,
    onChangeField: (field: string, value: any) => void
}

const usePetById = ({ id, onChangeField }: UsePetByNameProps) => {

    const [isPending, startTransition] = useTransition()
    const [petExists, setPetExists] = useState(false)

    const pets = useAppSelector(state => state.Pets.data)

    const onChangePet = useCallback((pet: IPet) => {
        startTransition(() => {
            onChangeField('pet.id', pet.id)
            onChangeField('pet.avatar', pet.avatar)
            onChangeField('pet.name', pet.name)
            onChangeField('pet.breed', pet.breed)
            onChangeField('pet.species', pet.species)
            onChangeField('pet.bloodType', pet.bloodType)
            onChangeField('pet.gender', pet.gender)
            onChangeField('pet.dateOfBirth', pet.dateOfBirth)

            setPetExists(true)
        })
    }, [onChangeField])

    useEffect(() => {
        if (!id) {
            return
        }
        const pet = pets?.find(pet => pet?.id === id)

        if (!pet) return

        onChangePet(pet)
    }, [pets, onChangePet, id])


    return {
        isPending,
        petExists
    }
}

export default usePetById