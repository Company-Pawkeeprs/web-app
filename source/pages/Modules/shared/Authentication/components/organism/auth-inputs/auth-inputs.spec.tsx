import { fireEvent, render, screen } from '@testing-library/react'
import { act } from 'react'
import { describe, expect, it, vi } from 'vitest'
import * as useAuth from '~/hooks/use-auth'
import ProviderClient from '~/store'
import AuthInputs from './auth-inputs'

const Wrapper = () => (
    <ProviderClient>
        <AuthInputs mode="tutor" />
    </ProviderClient>
)

describe('auth-inputs (Unit)', () => {
    it('deve corresponder ao snapshot', () => {
        const { container } = render(<Wrapper />)
        expect(container).toMatchSnapshot()
    })

    it('deve enviar o email em letra minúscula', () => {
        // Cria o spy para o useAuth
        const signInMock = vi.fn()
        vi.spyOn(useAuth, 'default').mockReturnValue({
            signIn: signInMock,
            password: 'mock@1Password',
            username: 'EMAIL@EXEMPLO.COM',
            isAuthenticated: false,
            isLoading: 1,
            onToggleRememberMe: () => {},
            rememberMe: false,
            user: null,
        })

        render(<Wrapper />)

        const usernameInput = screen.getByTestId('email-input')
        const passwordInput = screen.getByTestId('password-input')

        act(() => {
            fireEvent.change(usernameInput, {
                target: { value: 'EMAIL@EXEMPLO.COM' },
            })
            fireEvent.change(passwordInput, { target: { value: 'mock@1Password' } })
        })

        const submitButton = screen.getByTestId('submit-button')

        act(() => {
            fireEvent.submit(submitButton)
        })

        // Verifique se signIn foi chamado com o username em letras minúsculas
        expect(signInMock).toHaveBeenCalled()
    })

    it('deve renderizar o componente corretamente', () => {
        vi.spyOn(useAuth, 'default').mockReturnValue({
            signIn: vi.fn(),
            password: 'mockPassword',
            username: 'mockUsername',
            isAuthenticated: false,
            isLoading: 1,
            onToggleRememberMe: () => {},
            rememberMe: false,
            user: null,
        })

        render(<Wrapper />)

        expect(screen.getByText('Esqueceu a senha?')).toBeInTheDocument()
    })
})
