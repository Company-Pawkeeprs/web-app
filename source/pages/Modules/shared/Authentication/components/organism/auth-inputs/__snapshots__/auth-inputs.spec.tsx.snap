// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`auth-inputs (Unit) > deve corresponder ao snapshot 1`] = `
<div>
  <form
    action="#"
    class="w-full p-2"
  >
    <div
      class="w-full p-1"
    >
      <label
        class="text-xs font-semibold text-gray-500 gap-1 text-start items-center label justify-start pb-1 mb-0 h-fit"
        data-testid="label-username"
        for="username"
      >
        Email:
      </label>
      <div
        class="relative"
      >
        <input
          class="w-full px-4 py-2 duration-100 ease-in-out transition-shadow disabled:opacity-70 text-gray-600 disabled:bg-gray-200 text-base h-11 invalid:!shadow-outline-secondary-500 invalid:!text-secondary-500 invalid:border-secondary-500 invalid:focus:border-gray-300 border font-sans dark:bg-[#292e33] dark:text-gray-200 focus:ring-2 focus:ring-primary-500 bg-white focus:!border-none focus:!ring-2 focus:!ring-primary-500 outline-none rounded-md overflow-visible"
          ctx="[object Object]"
          data-testid="email-input"
          id="username"
          name="username"
          pattern="[^\\s]+"
          placeholder="Digite seu email"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="flex flex-col items-end justify-center w-full mb-3 position-relative"
    >
      <div
        class="w-full p-1"
      >
        <label
          class="text-xs font-semibold text-gray-500 gap-1 text-start items-center label justify-start pb-1 mb-0 h-fit"
          data-testid="label-password"
          for="password"
        >
          Senha:
        </label>
        <div
          class="relative"
        >
          <div
            class="absolute inset-y-0 flex items-center pl-1 text-sm text-gray-400 pointer-events-none left-1"
          >
            <svg
              aria-hidden="true"
              class="w-5 h-5 "
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.75 5.25a3 3 0 0 1 3 3m3 0a6 6 0 0 1-7.029 5.912c-.563-.097-1.159.026-1.563.43L10.5 17.25H8.25v2.25H6v2.25H2.25v-2.818c0-.597.237-1.17.659-1.591l6.499-6.499c.404-.404.527-1 .43-1.563A6 6 0 1 1 21.75 8.25Z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </div>
          <input
            class="w-full px-4 py-2 duration-100 ease-in-out transition-shadow disabled:opacity-70 text-gray-600 disabled:bg-gray-200 text-base h-11 invalid:!shadow-outline-secondary-500 invalid:!text-secondary-500 invalid:border-secondary-500 invalid:focus:border-gray-300 border font-sans dark:bg-[#292e33] dark:text-gray-200 focus:ring-2 focus:ring-primary-500 bg-white focus:!border-none focus:!ring-2 focus:!ring-primary-500 outline-none rounded-md overflow-visible !pl-8 !pr-8"
            ctx="[object Object]"
            data-testid="password-input"
            id="password"
            name="password"
            placeholder="Digite sua senha"
            position="right"
            type="password"
            value=""
          />
          <div
            class="absolute right-0 mr-2 transform -translate-y-1/2 top-1/2"
          >
            <button
              aria-label="toggle password visibility"
              class="flex items-center justify-center opacity-50"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="w-5 h-5"
                data-slot="icon"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3.98 8.223A10.477 10.477 0 0 0 1.934 12C3.226 16.338 7.244 19.5 12 19.5c.993 0 1.953-.138 2.863-.395M6.228 6.228A10.451 10.451 0 0 1 12 4.5c4.756 0 8.773 3.162 10.065 7.498a10.522 10.522 0 0 1-4.293 5.774M6.228 6.228 3 3m3.228 3.228 3.65 3.65m7.894 7.894L21 21m-3.228-3.228-3.65-3.65m0 0a3 3 0 1 0-4.243-4.243m4.242 4.242L9.88 9.88"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <a
        class="rounded-md text-center hover:brightness-90 transition duration-100 ease-in-out disabled:opacity-50 disabled:cursor-not-allowed bg-opacity-80 hover:bg-opacity-100 active:opacity-100 hover:transform hover:scale-[101%] disabled:hover:scale-100 text-opacity-100 text-secondary-500 dark:text-primary-600 hover:no-underline z-10 flex items-center content-end justify-end p-0 m-0 text-xs font-semibold h-fit w-fit"
        href="/forgot-password"
      >
        Esqueceu a senha?
      </a>
    </div>
    <button
      class="mobile:p-4 mobile:w-full w-40 flex-grow px-2 h-11 py-1 text-sm font-semibold rounded-md text-center leading-1 hover:brightness-90 transition duration-100 ease-in-out disabled:opacity-50 disabled:cursor-not-allowed bg-opacity-80 hover:bg-opacity-100 active:opacity-100 hover:transform hover:scale-[101%] disabled:hover:scale-100 text-opacity-100 text-gray-50 flex justify-center items-center bg-primary-500 !w-full"
      color="primary"
      data-testid="submit-button"
      disabled=""
      id=":r0:"
      type="submit"
    >
      Entrar
    </button>
  </form>
</div>
`;
